test/test-stylus-validity.c

```diff
@@ -0,0 +1,300 @@
+	   we have an eraser for each type, this will always be true, even if
+	   we forget to add the right eraser for a new pen.
+	 */
+	type = libwacom_stylus_get_type (stylus);
+	for (const WacomStylus **s = all_styli; *s; s++) {
+		if (libwacom_stylus_is_eraser(*s) &&
+		    libwacom_stylus_get_type(*s) == type) {
+			matching_eraser_found = TRUE;
+			break;
+		}
+	}
+
+	g_assert_true(matching_eraser_found);
+}
+
+static void
+test_puck(gconstpointer data)
+{
+	const WacomStylus *stylus = data;
+	int has_wheel = libwacom_stylus_has_wheel(stylus);
+	int has_lens = libwacom_stylus_has_lens(stylus);
+
+	/* 4D mouse is the only one with neither, everything
+	 * else has either wheel or lens */
+	if (libwacom_stylus_get_id(stylus) == 0x94) {
+		g_assert_false(has_wheel);
+		g_assert_false(has_lens);
+	} else {
+		g_assert_true(has_wheel != has_lens);
+	}
+}
+
+static void
+test_tilt(gconstpointer data)
+{
+	const WacomStylus *stylus = data;
+	gboolean has_tilt = libwacom_stylus_get_axes(stylus) & WACOM_AXIS_TYPE_TILT;
+	g_assert_true(has_tilt);
+}
+
+static void
+test_pressure(gconstpointer data)
+{
+	const WacomStylus *stylus = data;
+	gboolean has_pressure = libwacom_stylus_get_axes(stylus) & WACOM_AXIS_TYPE_PRESSURE;
+	g_assert_true(has_pressure);
+}
+
+static void
+test_distance(gconstpointer data)
+{
+	const WacomStylus *stylus = data;
+	gboolean has_distance = libwacom_stylus_get_axes(stylus) & WACOM_AXIS_TYPE_DISTANCE;
+	g_assert_true(has_distance);
+}
+
+static void
+test_name(gconstpointer data)
+{
+	const WacomStylus *stylus = data;
+
+	g_assert_nonnull(libwacom_stylus_get_name(stylus));
+}
+
+static void
+test_buttons(gconstpointer data)
+{
+	const WacomStylus *stylus = data;
+
+	g_assert_cmpint(libwacom_stylus_get_num_buttons(stylus), >, 0);
+}
+
+/* Wrapper function to make adding tests simpler. g_test requires
+ * a unique test case name so we assemble that from the test function and
+ * the stylus data.
+ */
+static inline void
+_add_test(const WacomStylus *stylus, GTestDataFunc func, const char *funcname)
+{
+	char buf[128];
+	const char *prefix;
+
+	/* tests must be test_foobar */
+	g_assert(strncmp(funcname, "test_", 5) == 0);
+	prefix = &funcname[5];
+
+	snprintf(buf, 128, "/stylus/%s/%03x-%s",
+		 prefix,
+		 libwacom_stylus_get_id(stylus),
+		 libwacom_stylus_get_name(stylus));
+	g_test_add_data_func(buf, stylus, func);
+}
+#define add_test(stylus, func_) \
+	_add_test(stylus, func_, #func_)
+
+static void
+setup_tests(const WacomStylus *stylus)
+{
+	add_test(stylus, test_name);
+	add_test(stylus, test_type);
+
+	if (libwacom_stylus_has_eraser(stylus))
+		add_test(stylus, test_eraser);
+
+	if (libwacom_stylus_get_type(stylus) == WSTYLUS_PUCK)
+		add_test(stylus, test_puck);
+
+	if (libwacom_stylus_is_eraser(stylus))
+		add_test(stylus, test_buttons);
+
+	if (libwacom_stylus_get_type(stylus) != WSTYLUS_PUCK) {
+		switch (libwacom_stylus_get_id(stylus)) {
+			case 0xffffd:
+			case 0x8e2:
+			case 0x862:
+				break;
```

Comment:
It should be more straightforward to check 0x8e2 and 0x862 for pressure/distance now that this is a switch statement. Don't have to if you don't want though.

Reply:
Going to leave that as follow-up, mostly because... huh? Sorry, didn't quite parse what you meant :)

Reply:
Included in #132